File: occasum-core/src/main/java/pw/stamina/occasum/node/visit/PropertyRegisterCleanupPropertyNodeVisitor.java
package pw.stamina.occasum.node.visit;

import pw.stamina.occasum.PropertyHandle;
import pw.stamina.occasum.node.PropertyNode;

import java.util.Objects;
import java.util.function.Consumer;

public final class PropertyRegisterCleanupPropertyNodeVisitor implements PropertyNodeVisitor {
    private final PropertyHandle handle;
    private final Consumer<PropertyNode> notifier;

    public PropertyRegisterCleanupPropertyNodeVisitor(PropertyHandle handle,
                                                      Consumer<PropertyNode> notifier) {
        Objects.requireNonNull(handle, "handle");
        Objects.requireNonNull(notifier, "notifier");

        this.handle = handle;
        this.notifier = notifier;
    }

    @Override
    public void visitNode(PropertyNode node) {
        if (node.getHandle() == handle) {
            assert node.hasParent();
            node.getParent().removeChild(node);
            notifier.accept(node);
        }
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/node/visit/PropertyRegisterCleanupPropertyNodeVisitor.kt
package pw.stamina.occasum.node.visit

import pw.stamina.occasum.PropertyHandle
import pw.stamina.occasum.node.PropertyNode

class PropertyRegisterCleanupPropertyNodeVisitor(
        private val handle: PropertyHandle,
        private val notifier: (PropertyNode) -> Unit)
    : PropertyNodeVisitor {

    override fun visitNode(node: PropertyNode) {
        if (node.handle === handle) {
            val parent = node.parent ?: return
            parent.removeChild(node)
            notifier(node)
        }
    }
}
